tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -2273,7 +2273,9 @@ void SpirvEmitter::doIfStmt(const IfStmt *ifStmt,
 }
 
 void SpirvEmitter::doReturnStmt(const ReturnStmt *stmt) {
-  if (const auto *retVal = stmt->getRetValue()) {
+  const auto *retVal = stmt->getRetValue();
+  bool returnsVoid = curFunction->getReturnType().getTypePtr()->isVoidType();
+  if (!returnsVoid && retVal) {
```

Comment:
Is it possible that a function does not return void and `retVal` is null?  If not, I would only check `returnsVoid` and assert that `retVal` is not null in the if-statement.

If it is possible, then you might need to handle that case as well.  You'll need to generate an `OpReturnValue` with an undef value for that case.  I don't think it is correct to have an `OpReturn` is a function with a non-void return type in spir-v.  See https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpReturn

Reply:
I checked and the frontend doesn't allow non-void functions to not return anything (`error: non-void function 'B' should return a value [-Wreturn-type`), so I switched to an assert. I'm not sure if there's a formal specification for HLSL – who should I talk to about HLSL structure?